nova/cmd/manage.py

```diff
@@ -2341,58 +2347,62 @@ def sync_aggregates(self, verbose=False):
-                                    '%(error)s') %
-                                  {'host': host, 'provider': rp_uuid,
-                                   'aggregate': aggregate_uuid,
-                                   'error': put_resp.text})
-                            return 3
-                elif resp.status_code == 404:
-                    # The resource provider wasn't found. Store this for later.
-                    providers_not_found[host] = rp_uuid
-                else:
-                    print(_('An error occurred getting resource provider '
-                            'aggregates from placement for provider '
-                            '%(provider)s. Error: %(error)s') %
-                          {'provider': rp_uuid, 'error': resp.text})
-                    return 2
+                for rp_uuid in rp_uuids:
+                    resp = placement.get(  # use 1.19 to get the generation
+                        '/resource_providers/%s/aggregates' % rp_uuid,
+                        version='1.19')
+                    if resp:
+                        body = resp.json()
+                        provider_aggregate_uuids = body['aggregates']
+                        # The moment of truth: is the provider in the same host
+                        # aggregate relationship?
+                        aggregate_uuid = aggregate.uuid
+                        if aggregate_uuid not in provider_aggregate_uuids:
+                            # Add the resource provider to this aggregate.
+                            provider_aggregate_uuids.append(aggregate_uuid)
+                            # Now update the provider aggregates using the
+                            # generation to ensure we're conflict-free.
+                            aggregate_update_body = {
+                                'aggregates': provider_aggregate_uuids,
+                                'resource_provider_generation':
+                                    body['resource_provider_generation']
+                            }
+                            put_resp = placement.put(
+                                '/resource_providers/%s/aggregates' % rp_uuid,
+                                aggregate_update_body, version='1.19')
+                            if put_resp:
+                                output(_('Successfully added host (%(host)s) '
+                                         'and provider (%(provider)s) to '
+                                         'aggregate (%(aggregate)s).') %
+                                       {'host': host, 'provider': rp_uuid,
+                                        'aggregate': aggregate_uuid})
+                            elif put_resp.status_code == 404:
+                                # We must have raced with a delete on the
+                                # resource provider.
+                                providers_not_found.setdefault(host, []) \
+                                    .append(rp_uuid)
```

Comment:
Why not a make it a defaultdict above? It would make this line and the one at the end simpler.

Reply:
(I didn't even know about `setdefault()` until now 🙂)

Reply:
no, thanks. `setdefault()` is fine.